Add RWKV-4

sgugger opened the pull request:
# What does this PR do?

This PR is a draft and while there is a working implementation of the model, there is still a lot to do :-)

This PR adds the RWKV model from [BlinkDL/RWKV-LM](https://github.com/BlinkDL/RWKV-LM) which is a RNN-like Transformers:  it has an attention layer and a feed-forward, but the attention is linear and can be expressed recurrently (more details coming in the doc page of the model).

Here is a code snippet to play with the model:
```py
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM

model = AutoModelForCausalLM.from_pretrained("sgugger/rwkv-7b-pile", torch_dtype=torch.float16, device_map="auto")
tokenizer = AutoTokenizer.from_pretrained("sgugger/rwkv-7b-pile")

prompt = "\nIn a shocking finding, scientist discovered a herd of dragons living in a remote, previously unexplored valley, in Tibet. Even more surprising to the researchers was the fact that the dragons spoke perfect Chinese."

inputs = tokenizer(prompt, return_tensors="pt").to(0)
output = model.generate(inputs["input_ids"], max_new_tokens=400, top_p=0.8, do_sample=True)
print(tokenizer.decode(output[0].tolist()))
```
To use the chat models (called raven):
```python
from transformers import AutoTokenizer, AutoModelForCausalLM, pipeline

model_id = "ybelkada/rwkv-raven-1b5"
model = AutoModelForCausalLM.from_pretrained(model_id).to(0)
tokenizer = AutoTokenizer.from_pretrained(model_id)

question = "Tell me about ravens"
prompt = f"### Instruction: {question}\n### Response:"

inputs = tokenizer(prompt, return_tensors="pt").to(0)
output = model.generate(inputs["input_ids"], max_new_tokens=100)

print(tokenizer.decode(output[0].tolist(), skip_special_tokens=True))
```

Fixes #20737 
Fixes #17230 

TODO:
- [x] Write documentation of the model explaining the linear attention and the recurrent formulas in the code
- [x] Make the model compatible with generate
- [x] Add output_attentions/output_hidden_states API
- [ ] Convert mode models and check conversion script is compatible
- [x] Tweak CUDA kernels for state to use the state for init
- [x] Make tests that pass
- [ ] Add attention mask to be able to batch sentences (might be in a followup PR)

cc @ArthurZucker and @younesbelkada

sgugger committed:
First draft of RWKV-4

sgugger committed:
Add support for generate

Blealtan commented on src/transformers/models/rwkv/modeling_rwkv.py:
```diff
@@ -0,0 +1,761 @@
+    # TODO: maybe jit, otherwise move inside forward
+    def extract_key_value(self, hidden, state=None):
+        # Mix hidden with the previous timestep to produce key, value, receptance
+        shifted = self.time_shift(hidden) if state is None or hidden.size(1) != 1 else state[1][:, :, self.layer_id]
```
Seems this shift mistakenly drops the previous `hidden` in `state` when provided with a sequence of length larger than 1. In the case of `state is not None and hidden.size(1) != 1`, it should cut the last token and prepend with the token from `state`.

sgugger replied:
Indeed, thanks for the pointer!

sgugger replied:
Should be fixed now!

YovaKem commented:
@sgugger thanks for doing this work, it's MUCH appreciated! Any chance you could share your script for converting a rwkv model to a HF checkpoint?

sgugger committed:
Style post-rebase

sgugger committed:
Properly use state

sgugger committed:
Write doc

sgugger committed:
Fix doc

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/transformers/pr_22797). All of your documentation changes will be reflected on that endpoint.

sgugger committed:
More math

sgugger committed:
Add model to README, dummies and clean config

sgugger committed:
Fix init

younesbelkada committed:
multiple fixes:

- fix common tests
- fix configuraion default values
- add CI test for checking state computation
- fix some CI tests

younesbelkada committed:
Merge branch 'main' into add_rwkv

younesbelkada committed:
correct tokenizer

younesbelkada committed:
Merge branch 'add_rwkv' of https://github.com/huggingface/transformers into add_rwkv

younesbelkada committed:
some tweaks

- fix config docstring
- fix failing tests

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add_rwkv

younesbelkada committed:
fix CI tests

- add output_attention / output_hidden_states
- override test_initialization
- fix failing CIs

younesbelkada committed:
fix conversion script

- fix sharded case
- add new arguments

younesbelkada committed:
add slow tests + more fixes on conversion script

younesbelkada committed:
add another test

younesbelkada committed:
final fixes

younesbelkada commented:
IMO the model is in a nice shape! Would love to have a round of review before I transfer the weights on the proper organization!

younesbelkada committed:
change single name variable

sgugger commented on tests/test_configuration_common.py:
```diff
@@ -93,15 +93,20 @@
 
 class ConfigTester(object):
-    def __init__(self, parent, config_class=None, has_text_modality=True, **kwargs):
+    def __init__(self, parent, config_class=None, has_text_modality=True, common_properties=None, **kwargs):
```
Thanks for the explanation!

younesbelkada committed:
add mock attention mask for pipeline to work

younesbelkada commented on src/transformers/models/rwkv/modeling_rwkv.py:
```diff
@@ -0,0 +1,796 @@
+    def forward(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        attention_mask: Optional[torch.LongTensor] = None,
```
I added a "mock" attention mask here so that `pipeline` won't fail complaining that the attention mask is outputted by the tokenizer and not used by the model. As we want to add the attention mask support anyway in the future, I thought it's the simplest solution now. To reproduce:
```bash
from transformers import pipeline

model_id = "ybelkada/rwkv-4-169m-pile"
prompt = "Hello"
pipe = pipeline("text-generation", model=model_id)
print(pipe(prompt, max_length=10))
```

younesbelkada committed:
correct eos token id

younesbelkada committed:
fix nits

younesbelkada committed:
add checkpoints

amyeroberts commented on src/transformers/models/rwkv/modeling_rwkv.py:
```diff
@@ -0,0 +1,805 @@
+            Optionally, instead of passing `input_ids` you can choose to directly pass an embedded representation. This
+            is useful if you want more control over how to convert `input_ids` indices into associated vectors than the
+            model's internal embedding lookup matrix.
+        state (tuple of five `torch.FloatTensor` of shape `(batch_size, hidden_size, num_hidden_layers)`, *optional*):
+            If passed along, the model uses the previous state in all the blocks (which will give the output for the
+            `input_ids` provided as if the model add `state_input_ids + input_ids` as context).
+        output_attentions (`bool`, *optional*):
+            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
+            tensors for more detail.
+        output_hidden_states (`bool`, *optional*):
+            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
+            more detail.
+        return_dict (`bool`, *optional*):
+            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
+        use_cache (`bool`, *optional*):
+            If set to `True`, the last state is returned and can be used to quickly generate the next logits.
```
ultra-nit: rearranging args here to match the order in the function signature 

```suggestion
        state (tuple of five `torch.FloatTensor` of shape `(batch_size, hidden_size, num_hidden_layers)`, *optional*):
            If passed along, the model uses the previous state in all the blocks (which will give the output for the
            `input_ids` provided as if the model add `state_input_ids + input_ids` as context).
        use_cache (`bool`, *optional*):
            If set to `True`, the last state is returned and can be used to quickly generate the next logits.
        output_attentions (`bool`, *optional*):
            Whether or not to return the attentions tensors of all attention layers. See `attentions` under returned
            tensors for more detail.
        output_hidden_states (`bool`, *optional*):
            Whether or not to return the hidden states of all layers. See `hidden_states` under returned tensors for
            more detail.
        return_dict (`bool`, *optional*):
            Whether or not to return a [`~utils.ModelOutput`] instead of a plain tuple.
```

amyeroberts reviewed (approved):
Thanks for adding this model - awesome PR 🔥  

Just some tiny nit comments from me :)

younesbelkada committed:
Apply suggestions from code review

Co-authored-by: amyeroberts <22614925+amyeroberts@users.noreply.github.com>

younesbelkada committed:
add `tie_word_embeddings` in docstring

younesbelkada commented on src/transformers/models/rwkv/configuration_rwkv.py:
```diff
@@ -0,0 +1,120 @@
+        rescale_every (`int`, *optional*, default to 6):
+            At inference, the hidden states (and weights of the correponding output layers) are divided by 2 every
+            `rescale_every` layer. If set to 0 or a negative number, no rescale is done.
+        use_cache (`bool`, *optional*, defaults to `True`):
```
Good point! added it in [f912a0d](https://github.com/huggingface/transformers/pull/22797/commits/f912a0d5f9143f86d5c3103707a17f2049586d1d)

younesbelkada commented on src/transformers/models/rwkv/modeling_rwkv.py:
```diff
@@ -0,0 +1,805 @@
+            ratio_0_to_1 = layer_id / (num_hidden_layers - 1)  # 0 to 1
+            ratio_1_to_almost0 = 1.0 - (layer_id / num_hidden_layers)  # 1 to ~0
+
+            ddd = torch.tensor(
```
I think it's a tensor that contains the expected initialized weights, I agree `ddd` is quite confusing :D 
changed it to something more explicit - `time_weight` as it corresponds to the initialization of `time_*` weights

sgugger replied:
Yeah didn't get any idea when converting the original code :sweat_smile:

younesbelkada committed:
change tensor name

younesbelkada reviewed (approved):
Thanks @sgugger for taking the lead on this! Learned a lot 🔥

BlinkDL commented:
@younesbelkada In README.md

The name should be "Bo Peng" (Peng is the surname) instead of "Peng Bo" :)

Blealtan commented on src/transformers/models/rwkv/modeling_rwkv.py:
```diff
@@ -0,0 +1,805 @@
+                forward_func = rwkv_cuda_kernel.forward_with_state_bf16
+            else:
+                forward_func = rwkv_cuda_kernel.forward_with_state
+            # TODO: update CUDA kernel so it uses the initial state provided here.
```
It seems this todo have been done?

Blealtan commented on src/transformers/models/rwkv/modeling_rwkv.py:
```diff
@@ -0,0 +1,805 @@
+
+    @staticmethod
+    # g stands for grad
+    def backward(ctx, g_output):
```
Any plan on supporting gradients on states? It would make chaining wkvs in training possible, getting rid of the seqlen limitation. It will also match the `_with_state` variant of WKV forward.

younesbelkada replied:
I think this makes sense, however I would advocate to do that in a follow up PR to at least unlock the model addition for users that already want to use the model with `transformers`

Blealtan replied:
May I help on that later?

younesbelkada replied:
Sure ! You are more than welcome to help us on that

younesbelkada committed:
fix final nits

Blealtan commented on src/transformers/models/rwkv/modeling_rwkv.py:
```diff
@@ -0,0 +1,804 @@
+        time_decay, time_first, key, value, output = ctx.saved_tensors
+        # The CUDA kernel will fill those tensors.
+        g_time_decay = torch.empty_like(
+            time_decay,
```
If I read it right, `time_decay`/`time_first` is of shape `(C, )` while the CUDA kernel requires `gw` and `gu` of shape `(B, C)` (see `wkv_cuda.cu:140-141`). It may cause VRAM overflow as well as wrong results. I didn't set up the environment to do the test, but I suspect the current `g_time_decay/first` after the summation in lines 192-193 will become scalars, which is unexpected.

younesbelkada committed:
Merge remote-tracking branch 'upstream/main' into add_rwkv

sgugger committed:
Trigger CI

YovaKem commented:
hi @sgugger, thanks A TON for this merge! I am trying to train a new model of type and facing the following error: 
```
Traceback (most recent call last):
  File "train.py", line 229, in <module>
    main(model_args, data_args, training_args)
  File "train.py", line 193, in main
    trainer.train()
  File "transformers/src/transformers/trainer.py", line 1664, in train
    return inner_training_loop(
  File "transformers/src/transformers/trainer.py", line 1940, in _inner_training_loop
    tr_loss_step = self.training_step(model, inputs)
  File "transformers/src/transformers/trainer.py", line 2753, in training_step
    loss.backward()
  File ".conda/envs/rwkv-eval-3.9/lib/python3.9/site-packages/torch/_tensor.py", line 487, in backward
    torch.autograd.backward(
  File ".conda/envs/rwkv-eval-3.9/lib/python3.9/site-packages/torch/autograd/__init__.py", line 200, in backward
    Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
  File ".conda/envs/rwkv-eval-3.9/lib/python3.9/site-packages/torch/autograd/function.py", line 274, in apply
    return user_fn(self, *args)
TypeError: backward() takes 2 positional arguments but 3 were given
```

Any idea how I should go about fixing this? Thanks!

YovaKem commented:
I managed to get the code to run with some changes to the forward() and backward() functions:

```
class RwkvLinearAttention(torch.autograd.Function):
    @staticmethod
    def forward(ctx, time_decay, time_first, key, value, state=None, return_state=False):

        batch_size, seq_len, hidden_size = key.size()
        if seq_len > rwkv_cuda_kernel.max_seq_length:
            raise ValueError(
                f"Cannot process a batch with {seq_len} tokens at the same time, use a maximum of "
                f"{rwkv_cuda_kernel.max_seq_length} with this model."
            )
        if batch_size * hidden_size % min(hidden_size, 32) != 0:
            raise ValueError(
                f"The product of batch size ({batch_size}) and hidden size ({hidden_size}) needs to be a round "
                f"multiple of {min(hidden_size, 32)}."
            )

        ctx.input_dtype = key.dtype

        if (
            time_decay.device.type != "cuda"
            or time_first.device.type != "cuda"
            or key.device.type != "cuda"
            or value.device.type != "cuda"
        ):
            raise ValueError("Calling the CUDA kernel for wkv attention requires all tensors to be on CUDA devices.")

        time_decay = -torch.exp(time_decay.float().contiguous())
        if key.dtype == torch.float16:
            time_first = time_first.float()
            key = key.float()
            value = value.float()
        time_first = time_first.contiguous()
        key = key.contiguous()
        value = value.contiguous()
        # The CUDA kernel will fill this tensor.
        output = torch.empty_like(key, memory_format=torch.contiguous_format)
        if return_state or state is not None:
            if state is None:
                state = torch.zeros(
                    batch_size,
                    hidden_size,
                    3,
                    dtype=torch.float32,
                    device=key.device,
                    memory_format=torch.contiguous_format,
                )
                state[:, :, 2] -= 1e38
            else:
                state = torch.cat([s.unsqueeze(2) for s in state], dim=2).contiguous()

            if key.dtype == torch.bfloat16:
                forward_func = rwkv_cuda_kernel.forward_with_state_bf16
            else:
                forward_func = rwkv_cuda_kernel.forward_with_state
            forward_func(time_decay, time_first.to(key.dtype), key, value, output, state)
        else:
            forward_func = rwkv_cuda_kernel.forward_bf16 if key.dtype == torch.bfloat16 else rwkv_cuda_kernel.forward
            forward_func(time_decay, time_first.to(key.dtype), key, value, output)
        ctx.save_for_backward(time_decay, time_first, key, value, output)

        if state is not None:
            state = [s.squeeze(2) for s in torch.chunk(state, 3, dim=2)]

        return output.to(ctx.input_dtype), state
```

```
    def backward(ctx, g_output, g_state):
        input_dtype = ctx.input_dtype

        time_decay, time_first, key, value, output = ctx.saved_tensors
        # The CUDA kernel will fill those tensors.
        g_time_decay = torch.empty_like(
            time_decay,
            memory_format=torch.contiguous_format,
            dtype=torch.bfloat16 if input_dtype == torch.bfloat16 else torch.float32,
        )
        g_time_first = torch.empty_like(
                time_first,
                memory_format=torch.contiguous_format,
                dtype=torch.bfloat16 if input_dtype == torch.bfloat16 else torch.float32,
        )
        g_key = torch.empty_like(key, memory_format=torch.contiguous_format)
        g_value = torch.empty_like(value, memory_format=torch.contiguous_format)

        if input_dtype == torch.float16:
            g_output = g_output.float()
        backward_func = rwkv_cuda_kernel.backward_bf16 if input_dtype == torch.bfloat16 else rwkv_cuda_kernel.backward
        backward_func(
            time_decay,
            time_first.to(key.dtype),
            key,
            value,
            output,
            g_output.contiguous(),
            g_time_decay,
            g_time_first,
            g_key,
           g_value,
        )
        #g_time_decay = torch.sum(g_time_decay, dim=0)
        #g_time_first = torch.sum(g_time_first, dim=0)

        return (
            g_time_decay.to(input_dtype),
            g_time_first.to(input_dtype),
            g_key.to(input_dtype),
            g_value.to(input_dtype),
            None,
            None
        )
```

One problem I run into now is that although I'm trying to train a fairly small model (12 layers, 256 hidden size, 64 context size) I can only train with a very small batch size (16) on a 40GB A100 card. For comparison, a RoBERTa model with a similar size allows for a bs of 256. This seems counterintuitive to me, but I might be wrong.

Another issue I observed is instability: in some cases, within the first 3 steps of training the loss goes from something normal like 10 to 90543067814198.3 and then to 0.0. This seems to happen more when bf16 training is disabled and at higher batch sizes when bf16 training is enabled.

Blealtan commented:
@YovaKem Would you mind try change this

```python
# The CUDA kernel will fill those tensors.
g_time_decay = torch.empty_like(
    time_decay,
    memory_format=torch.contiguous_format,
    dtype=torch.bfloat16 if input_dtype == torch.bfloat16 else torch.float32,
)
g_time_first = torch.empty_like(time_first, memory_format=torch.contiguous_format)
```

to

```python
# The CUDA kernel will fill those tensors.
g_time_decay = torch.empty(
    k.shape[0], k.shape[2],
    memory_format=torch.contiguous_format,
    dtype=torch.bfloat16 if input_dtype == torch.bfloat16 else torch.float32,
)
g_time_first = torch.empty(k.shape[0], k.shape[2], memory_format=torch.contiguous_format)
```

I suspect there's an overflow in the current code, as mentioned above in the review comment but not tested yet. The binary distribution on PyPI does not include the cuda kernels XD

YovaKem commented:
Thanks @Blealtan! I guess you meant `k` for `key`? I added bf16 support for `g_time_first` (I get an error otherwise) and put the tensors on CUDA

```python
        # The CUDA kernel will fill those tensors.
        g_time_decay = torch.empty(
            key.shape[0], key.shape[2],
            memory_format=torch.contiguous_format,
            dtype=torch.bfloat16 if input_dtype == torch.bfloat16 else torch.float32,
        ).to(key.device)
        g_time_first = torch.empty(
                key.shape[0], key.shape[2],
                memory_format=torch.contiguous_format,
                dtype=torch.bfloat16 if input_dtype == torch.bfloat16 else torch.float32,
        ).to(key.device)
```

This seems to solve both the OOM issue and the instability!

One question re your comment of state gradients - I now saw this

> It will also match the _with_state variant of WKV forward.

In what cases is the _with_state variant used? As far as I can see the model I'm training is not passing states at all during the forward step.

lambdaofgod commented:
Hey @sgugger how did you prepare the models? Could you point us how to convert original .pth or .safetensors model to your format? Thanks!

PS
Awesome RWKV joined transformers!

3outeille commented:
Hi, I tried to evaluate perplexity on wikitext2 using HuggingFace RWKV but found a weird behavior (gist to reproduce the bug: https://gist.github.com/3outeille/e74ec833ec2800a94325f8dad8e0da3d).

- When model is fully loaded on CPU or GPU, perlexity is fine
- When some block of RWKV are loaded in CPU and GPU, perplexity is high 

Any idea ?

Blealtan commented:
@YovaKem AFAIK, `with_state` is used only in inference now (in existing non-`transformers` implementations throughout the RWKV community). However, with proper implementation, this will allow more efficient training on long sequences, but it has not yet been implemented.

sgugger commented:
Normally custom kernels should be included in 4.29.2, sorry for the inconvenience. We added stronger to checks to make sure they don't disappear again in a future release.

Wednesday657 commented:
Hi, can i ask a simple question about RWKV kernel? The rwkv model without customized kernel uses a `for loop` here:
https://github.com/huggingface/transformers/blob/3658488ff77ff8d45101293e749263acf437f4d5/src/transformers/models/rwkv/modeling_rwkv.py#L223-L241

I am not familiar with cuda kernel. So i am not sure whether the customized cuda kernel still computes sequentially and delivers a faster `for loop`, or just make the computation parallelized in GPU?

fullstackwebdev commented:
Putting this here so it doesn't get lost. 

I am trying to run microsoft guidance (https://github.com/microsoft/guidance) on RWKV through transformers and I am getting an error

`AttributeError: 'RwkvCausalLMOutput' object has no attribute 'past_key_values'`

which can be reproduced here: https://gist.github.com/fullstackwebdev/a6523374e6687825fcb92ca74048c12b

younesbelkada commented:
@fullstackwebdev 
I don't think the fix should go inside `transformers` as this means we should always output `past_key_values=None` - which is quite misleading as by desing RWKV does not rely on `past_key_values` for caching - as the tokens are processed one by one. I made https://github.com/microsoft/guidance/pull/91 that fixed the issue in my local env
